lib/puppet/settings.rb

```diff
@@ -1156,11 +1156,23 @@ def add_environment_resources(catalog, sections)
     if configured_environment == "production" && envdir && Puppet::FileSystem.exist?(envdir)
       configured_environment_path = File.join(envdir, configured_environment)
       if !Puppet::FileSystem.symlink?(configured_environment_path)
-        catalog.add_resource(
-          Puppet::Resource.new(:file,
-                               configured_environment_path,
-                               :parameters => { :ensure => 'directory' })
-        )
+        eid = Process::UID.eid
+        if eid == 0 or eid == Etc.getpwnam(self[:user]).uid
+          catalog.add_resource(
+            Puppet::Resource.new(:file,
+                                 configured_environment_path,
+                                 :parameters => { :ensure => 'directory',
+                                                  :owner => self[:user],
+                                                  :group => self[:group],
+                                                  :mode => "0750" })
```

Comment:
I think you want to set `owner` and `group` to `service`. If puppet is running as `root` and the service user/group account exists, then puppet will set the owner/group accordingly.

If we're not running as root, then puppet can't chown, so no point in trying.

If we are root, but the `puppet` user/group doesn't exist, then we fall back to either root or a platform-dependent account.

Note the `puppet` user/group accounts will typically only exist when the agent is running on a host that is also a puppetmaster, or the agent upgraded from 3.x.